specs/phase1/beacon-chain.md

```diff
@@ -767,236 +757,176 @@ def validate_attestation(state: BeaconState, attestation: Attestation) -> None:
 
     committee = get_beacon_committee(state, data.slot, data.index)
     assert len(attestation.aggregation_bits) == len(committee)
-
-    if data.target.epoch == get_current_epoch(state):
-        assert data.source == state.current_justified_checkpoint
-    else:
-        assert data.source == state.previous_justified_checkpoint
-
-    # Type 1: on-time attestations
-    if is_on_time_attestation(state, data):
-        # Correct parent block root
-        assert data.beacon_block_root == get_block_root_at_slot(state, compute_previous_slot(state.slot))
-        # Correct shard number
-        shard = compute_shard_from_committee_index(state, data.index, data.slot)
-        assert data.shard == shard
-        # NOTE: We currently set `PHASE_1_FORK_SLOT` to `GENESIS_SLOT` for test vectors.
-        if data.slot > GENESIS_SLOT:
-            # On-time attestations should have a non-empty shard transition root
-            assert data.shard_transition_root != hash_tree_root(ShardTransition())
-        else:
-            assert data.shard_transition_root == hash_tree_root(ShardTransition())
-    # Type 2: no shard transition
-    else:
-        # Ensure delayed attestation
-        assert data.slot < compute_previous_slot(state.slot)
-        # Late attestations cannot have a shard transition root
-        assert data.shard_transition_root == Root()
-
-    # Signature check
-    assert is_valid_indexed_attestation(state, get_indexed_attestation(state, attestation))
-```
-
-###### Updated `process_attestation`
-
-```python
-def process_attestation(state: BeaconState, attestation: Attestation) -> None:
-    validate_attestation(state, attestation)
-    # Store pending attestation for epoch processing
+    
     pending_attestation = PendingAttestation(
         aggregation_bits=attestation.aggregation_bits,
         data=attestation.data,
         inclusion_delay=state.slot - attestation.data.slot,
         proposer_index=get_beacon_proposer_index(state),
-        crosslink_success=False,  # To be filled in during process_shard_transitions
     )
-    if attestation.data.target.epoch == get_current_epoch(state):
+
+    if data.target.epoch == get_current_epoch(state):
+        assert data.source == state.current_justified_checkpoint
         state.current_epoch_attestations.append(pending_attestation)
+        shard_transition_candidates = state.current_shard_transition_candidates
     else:
+        assert data.source == state.previous_justified_checkpoint
         state.previous_epoch_attestations.append(pending_attestation)
+        shard_transition_candidates = state.previous_shard_transition_candidates
+
+    # Signature check
+    assert is_valid_indexed_attestation(state, get_indexed_attestation(state, attestation))
+
+    # Process shard transition
+    candidate_exists = False
+    for candidate in shard_transition_candidates:
```

Comment:
Would it be better to construct `current_shard_transition_candidates` and `previous_shard_transition_candidates` as nested lists. Where the indices corresponds with slot in epoch.

```[[transitions from start slot],[transitions from start slot+1],[transitions from start slot+2]...]```

Reply:
hmm what is the value of dividing by slot? fast search?

Reply:
Faster lookup and insertion to OR bitfields. Although i dont think the complexity is worth it, i can write some benchmarks for this later